### PR TITLE
Change ManuallyDrop<T> to a lang item.

### LINKED ISSUE #47034: ManuallyDrop<T> where T: ?Sized
`ManuallyDrop<T>` currently requires `T: Sized`. It would be nice if we could remove this restriction. I'm not sure how to do it with the current implementation using `union` though.

### PR DESCRIPTION
**Pull request by eddyb**
This PR implements the approach @RalfJung proposes in https://internals.rust-lang.org/t/pre-rfc-unions-drop-types-and-manuallydrop/8025 (lang item `struct` instead of `union`).

A followup PR can easily solve #47034 as well, by just adding a few `?Sized` to `libcore/mem.rs`.

r? @nikomatsakis

### CODE REVIEW

**Comment by RalfJung** [comment]
> `ManuallySized<T>`

;) Should be `ManuallyDrop<T>`

File: src/libcore/mem.rs [hunk 1]
```diff
@@ -954,11 +953,18 @@ pub fn discriminant<T>(v: &T) -> Discriminant<T> {
 ///     }
 /// }
 /// ```
+#[cfg(not(stage0))]
 #[stable(feature = "manually_drop", since = "1.20.0")]
-#[allow(unions_with_drop_fields)]
-#[derive(Copy)]
-pub union ManuallyDrop<T>{ value: T }
+#[lang = "manually_drop"]
+#[derive(Copy, Clone, Debug, Default, PartialEq, Eq, PartialOrd, Ord, Hash)]
```
Comment (scottmcm) [inline src/libcore/mem.rs]: Nice that these can all just be derives, now! :+1:

**Comment by nikomatsakis** [comment]
@eddyb 

```
[01:16:03] reference/destructors.html:217: broken link - std/mem/union.ManuallyDrop.html
[01:16:03] reference/print.html:6644: broken link - std/mem/union.ManuallyDrop.html
```

r=me once travis is happy

File: src/librustc/traits/query/dropck_outlives.rs [hunk 2]
```diff
@@ -243,7 +243,10 @@ pub fn trivial_dropck_outlives<'tcx>(tcx: TyCtxt<'_, '_, 'tcx>, ty: Ty<'tcx>) ->
 
         ty::TyAdt(def, _) => {
             if def.is_union() {
-                // Unions never run have a dtor.
+                // Unions never have a dtor.
+                true
```
Comment (RalfJung) [inline src/librustc/traits/query/dropck_outlives.rs]: A union implementing `Drop` itself is not a concern here?

**Comment by RalfJung** [comment]
This makes `ManuallyDrop` for unsized types insta-stable. Does that require an FCP?

@eddyb said he doesn't have much time to follow up on this, but I think I can take over. I should probably also add at least one test because it seems there are none?

File: src/librustc/traits/query/dropck_outlives.rs (hunk 2 above)
Comment (eddyb) [reply src/librustc/traits/query/dropck_outlives.rs]: Good question, cc @arielb1 @pnkfelix @nikomatsakis

**Comment by eddyb** [comment]
FWIW we would only need an FCP for the second commit, the first one can be landed independently and then union-related changes can be made in parallel with generalized to `!Sized`.

**Comment by RalfJung** [comment]
So shall we split the PR?

**Comment by RalfJung** [comment]
PR submitted to fix link in reference: https://github.com/rust-lang-nursery/reference/pull/373

**Comment by RalfJung** [comment]
@TimNN Bot shows empty log.

**Comment by RalfJung** [comment]
Updated reference submodule so the links should be fixed.

**Comment by RalfJung** [comment]
Dang, unrelated updates in the reference contain broken links.

**Comment by RalfJung** [comment]
> r=me once travis is happy

@bors r=nikomatsakis

File: src/librustc/traits/query/dropck_outlives.rs (hunk 2 above)
Comment (RalfJung) [reply src/librustc/traits/query/dropck_outlives.rs]: Actually, yes, I think it is a concern. :D  https://github.com/rust-lang/rust/issues/52786
